File: src/output.ts
/**
 * Output formatting utilities
 */

import type { ToolInfo } from './client.js';
import type { ServerConfig } from './config.js';
import { isHttpServer } from './config.js';

// ANSI color codes
const colors = {
  reset: '\x1b[0m',
  bold: '\x1b[1m',
  dim: '\x1b[2m',
  cyan: '\x1b[36m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  magenta: '\x1b[35m',
};

/**
 * Check if output should be colorized
 */
function shouldColorize(): boolean {
  return process.stdout.isTTY && !process.env.NO_COLOR;
}

/**
 * Apply color if terminal supports it
 */
function color(text: string, colorCode: string): string {
  if (!shouldColorize()) return text;
  return `${colorCode}${text}${colors.reset}`;
}

/**
 * Format server list for display
 */
export function formatServerList(
  servers: Array<{ name: string; tools: ToolInfo[]; instructions?: string }>,
  withDescriptions: boolean,
): string {
  const lines: string[] = [];

  for (const server of servers) {
    lines.push(color(server.name, colors.bold + colors.cyan));

    // Show instructions if available (first line only in list view, or all if short)
    if (server.instructions) {
      const instructionLines = server.instructions.split('\n');
      const firstLine = instructionLines[0].slice(0, 100);
      const suffix = instructionLines.length > 1 || instructionLines[0].length > 100 ? '...' : '';
      lines.push(`  ${color(`Instructions: ${firstLine}${suffix}`, colors.dim)}`);
    }

    for (const tool of server.tools) {
      if (withDescriptions && tool.description) {
        lines.push(`  • ${tool.name} - ${color(tool.description, colors.dim)}`);
      } else {
        lines.push(`  • ${tool.name}`);
      }
    }

    lines.push(''); // Empty line between servers
  }

  return lines.join('\n').trimEnd();
}

/**
 * Format search results
 */
export function formatSearchResults(
  results: Array<{ server: string; tool: ToolInfo }>,
  withDescriptions: boolean,
): string {
  const lines: string[] = [];

  for (const result of results) {
    const server = color(result.server, colors.cyan);
    const tool = color(result.tool.name, colors.green);
    // Always show description if available (grep is for discovery)
    if (result.tool.description) {
      lines.push(`${server} ${tool} ${color(result.tool.description, colors.dim)}`);
    } else {
      lines.push(`${server} ${tool}`);
    }
  }

  return lines.join('\n');
}

/**
 * Format server details
 */
export function formatServerDetails(
  serverName: string,
  config: ServerConfig,
  tools: ToolInfo[],
  withDescriptions = false,
  instructions?: string,
): string {
  const lines: string[] = [];

  lines.push(
    `${color('Server:', colors.bold)} ${color(serverName, colors.cyan)}`,
  );

  if (isHttpServer(config)) {
    lines.push(`${color('Transport:', colors.bold)} HTTP`);
    lines.push(`${color('URL:', colors.bold)} ${config.url}`);
  } else {
    lines.push(`${color('Transport:', colors.bold)} stdio`);
    lines.push(
      `${color('Command:', colors.bold)} ${config.command} ${(config.args || []).join(' ')}`,
    );
  }

  if (instructions) {
    lines.push('');
    lines.push(`${color('Instructions:', colors.bold)}`);
    // Indent multi-line instructions
    const indentedInstructions = instructions
      .split('\n')
      .map(line => `  ${line}`)
      .join('\n');
    lines.push(indentedInstructions);
  }

  lines.push('');
  lines.push(`${color(`Tools (${tools.length}):`, colors.bold)}`);

  for (const tool of tools) {
    lines.push(`  ${color(tool.name, colors.green)}`);
    if (withDescriptions && tool.description) {
      lines.push(`    ${color(tool.description, colors.dim)}`);
    }

    // Show parameters from schema
    const schema = tool.inputSchema as {
      properties?: Record<string, { type?: string; description?: string }>;
      required?: string[];
    };
    if (schema.properties) {
      lines.push(`    ${color('Parameters:', colors.yellow)}`);
      for (const [name, prop] of Object.entries(schema.properties)) {
        const required = schema.required?.includes(name)
          ? 'required'
          : 'optional';
        const type = prop.type || 'any';
        const desc =
          withDescriptions && prop.description ? ` - ${prop.description}` : '';
        lines.push(`      • ${name} (${type}, ${required})${desc}`);
      }
    }
    lines.push('');
  }

  return lines.join('\n').trimEnd();
}

/**
 * Format tool schema
 */
export function formatToolSchema(serverName: string, tool: ToolInfo): string {
  const lines: string[] = [];

  lines.push(
    `${color('Tool:', colors.bold)} ${color(tool.name, colors.green)}`,
  );
  lines.push(
    `${color('Server:', colors.bold)} ${color(serverName, colors.cyan)}`,
  );
  lines.push('');

  if (tool.description) {
    lines.push(`${color('Description:', colors.bold)}`);
    lines.push(`  ${tool.description}`);
    lines.push('');
  }

  lines.push(`${color('Input Schema:', colors.bold)}`);
  lines.push(JSON.stringify(tool.inputSchema, null, 2));

  return lines.join('\n');
}

/**
 * Format tool call result
 */
export function formatToolResult(result: unknown): string {
  if (typeof result === 'object' && result !== null) {
    const r = result as { content?: Array<{ type: string; text?: string }> };

    // Handle MCP tool result format
    if (r.content && Array.isArray(r.content)) {
      const textParts = r.content
        .filter((c) => c.type === 'text' && c.text)
        .map((c) => c.text);

      if (textParts.length > 0) {
        return textParts.join('\n');
      }
    }
  }

  // Fallback to JSON
  return JSON.stringify(result, null, 2);
}

/**
 * Format as JSON
 */
export function formatJson(data: unknown): string {
  return JSON.stringify(data, null, 2);
}

/**
 * Format error message
 */
export function formatError(message: string): string {
  return color(`Error: ${message}`, '\x1b[31m'); // Red
}

File: src/output.rs
use crate::config::ServerConfig;
use std::io::IsTerminal;

// ANSI color codes
const RESET: &str = "\x1b[0m";
const BOLD: &str = "\x1b[1m";
const DIM: &str = "\x1b[2m";
const CYAN: &str = "\x1b[36m";
const GREEN: &str = "\x1b[32m";
const YELLOW: &str = "\x1b[33m";

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct ToolInfo {
    pub name: String,
    pub description: Option<String>,
    pub input_schema: serde_json::Value,
}

pub struct ServerInfo {
    pub name: String,
    pub tools: Vec<ToolInfo>,
    pub instructions: Option<String>,
}

fn should_colorize() -> bool {
    std::io::stdout().is_terminal() && std::env::var("NO_COLOR").is_err()
}

fn color(text: &str, color_code: &str) -> String {
    if !should_colorize() {
        return text.to_string();
    }
    format!("{}{}{}", color_code, text, RESET)
}

pub fn format_server_list(servers: &[ServerInfo], with_descriptions: bool) -> String {
    let mut lines = Vec::new();

    for server in servers {
        lines.push(color(&server.name, &format!("{}{}", BOLD, CYAN)));

        if let Some(ref inst) = server.instructions {
            let inst_lines: Vec<&str> = inst.split('\n').collect();
            let first_line = if inst_lines[0].len() > 100 {
                &inst_lines[0][..100]
            } else {
                inst_lines[0]
            };
            let suffix = if inst_lines.len() > 1 || inst_lines[0].len() > 100 {
                "..."
            } else {
                ""
            };
            lines.push(format!(
                "  {}",
                color(&format!("Instructions: {}{}", first_line, suffix), DIM)
            ));
        }

        for tool in &server.tools {
            if with_descriptions {
                if let Some(ref desc) = tool.description {
                    lines.push(format!("  • {} - {}", tool.name, color(desc, DIM)));
                } else {
                    lines.push(format!("  • {}", tool.name));
                }
            } else {
                lines.push(format!("  • {}", tool.name));
            }
        }

        lines.push(String::new()); // Empty line between servers
    }

    let joined = lines.join("\n");
    joined.trim_end().to_string()
}

pub struct SearchResult {
    pub server: String,
    pub tool: ToolInfo,
}

pub fn format_search_results(results: &[SearchResult], _with_descriptions: bool) -> String {
    let mut lines = Vec::new();

    for result in results {
        let server_colored = color(&result.server, CYAN);
        let tool_colored = color(&result.tool.name, GREEN);

        if let Some(ref desc) = result.tool.description {
            lines.push(format!(
                "{} {} {}",
                server_colored,
                tool_colored,
                color(desc, DIM)
            ));
        } else {
            lines.push(format!("{} {}", server_colored, tool_colored));
        }
    }

    lines.join("\n")
}

pub fn format_server_details(
    server_name: &str,
    config: &ServerConfig,
    tools: &[ToolInfo],
    with_descriptions: bool,
    instructions: Option<&str>,
) -> String {
    let mut lines = Vec::new();

    lines.push(format!("{}: {}", color("Server", BOLD), color(server_name, CYAN)));

    match config {
        ServerConfig::Http(hc) => {
            lines.push(format!("{}: HTTP", color("Transport", BOLD)));
            lines.push(format!("{}: {}", color("URL", BOLD), hc.url));
        }
        ServerConfig::Stdio(sc) => {
            lines.push(format!("{}: stdio", color("Transport", BOLD)));
            let args_str = sc.args.as_ref().map(|a| a.join(" ")).unwrap_or_default();
            lines.push(format!(
                "{}: {} {}",
                color("Command", BOLD),
                sc.command,
                args_str
            ));
        }
    }

    if let Some(inst) = instructions {
        lines.push(String::new());
        lines.push(color("Instructions:", BOLD));
        let indented = inst
            .split('\n')
            .map(|l| format!("  {}", l))
            .collect::<Vec<_>>()
            .join("\n");
        lines.push(indented);
    }

    lines.push(String::new());
    lines.push(format!("{}:", color(&format!("Tools ({})", tools.len()), BOLD)));

    for tool in tools {
        lines.push(format!("  {}", color(&tool.name, GREEN)));
        if with_descriptions {
            if let Some(ref desc) = tool.description {
                lines.push(format!("    {}", color(desc, DIM)));
            }
        }

        // Show parameters from schema
        if let Some(obj) = tool.input_schema.as_object() {
            if let Some(properties) = obj.get("properties").and_then(|p| p.as_object()) {
                lines.push(format!("    {}", color("Parameters:", YELLOW)));
                let required_list: Vec<&str> = obj
                    .get("required")
                    .and_then(|r| r.as_array())
                    .map(|arr| {
                        arr.iter()
                            .filter_map(|val| val.as_str())
                            .collect::<Vec<_>>()
                    })
                    .unwrap_or_default();

                for (name, prop) in properties {
                    let required = if required_list.contains(&name.as_str()) {
                        "required"
                    } else {
                        "optional"
                    };
                    let ty = prop
                        .get("type")
                        .and_then(|t| t.as_str())
                        .unwrap_or("any");
                    let desc = if with_descriptions {
                        prop.get("description")
                            .and_then(|d| d.as_str())
                            .map(|d| format!(" - {}", d))
                            .unwrap_or_default()
                    } else {
                        String::new()
                    };
                    lines.push(format!(
                        "      • {} ({}, {}){}",
                        name, ty, required, desc
                    ));
                }
            }
        }
        lines.push(String::new());
    }

    lines.join("\n").trim_end().to_string()
}

pub fn format_tool_schema(server_name: &str, tool: &ToolInfo) -> String {
    let mut lines = Vec::new();

    lines.push(format!("{}: {}", color("Tool", BOLD), color(&tool.name, GREEN)));
    lines.push(format!("{}: {}", color("Server", BOLD), color(server_name, CYAN)));
    lines.push(String::new());

    if let Some(ref desc) = tool.description {
        lines.push(color("Description:", BOLD));
        lines.push(format!("  {}", desc));
        lines.push(String::new());
    }

    lines.push(color("Input Schema:", BOLD));
    lines.push(serde_json::to_string_pretty(&tool.input_schema).unwrap_or_default());

    lines.join("\n")
}

pub fn format_tool_result(result: &serde_json::Value) -> String {
    // Replicates TS formatToolResult:
    // Extract text content from MCP content array if present
    if let Some(obj) = result.as_object() {
        if let Some(content) = obj.get("content").and_then(|c| c.as_array()) {
            let text_parts: Vec<&str> = content
                .iter()
                .filter(|c| c.get("type").and_then(|t| t.as_str()) == Some("text"))
                .filter_map(|c| c.get("text").and_then(|t| t.as_str()))
                .collect();

            if !text_parts.is_empty() {
                return text_parts.join("\n");
            }
        }
    }

    serde_json::to_string_pretty(result).unwrap_or_default()
}

pub fn format_json(data: &serde_json::Value) -> String {
    serde_json::to_string_pretty(data).unwrap_or_default()
}

pub fn format_error(message: &str) -> String {
    color(&format!("Error: {}", message), "\x1b[31m")
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn test_format_tool_result_with_text() {
        let payload = json!({
            "content": [
                {
                    "type": "text",
                    "text": "Hello world from tool"
                }
            ]
        });
        let formatted = format_tool_result(&payload);
        assert_eq!(formatted, "Hello world from tool");
    }

    #[test]
    fn test_format_tool_result_raw_json() {
        let payload = json!({
            "some_key": "some_value"
        });
        let formatted = format_tool_result(&payload);
        assert!(formatted.contains("\"some_key\": \"some_value\""));
    }

    #[test]
    fn test_format_error() {
        // Since test env might have different terminal flags, check result contains "Error: "
        let formatted = format_error("Something went wrong");
        assert!(formatted.contains("Error: Something went wrong"));
    }
}
